# Role-Playing: Consulting Firm and Client Case

**Note:** You might need to install new libraries to run this file.

## The background: 

For this activity, you will participate in a role-playing business case involving a consulting firm and a client. 

The client is **GLOBAL MANUFACTURING SOLUTIONS**, a multinational company seeking to enhance its logistics and transportation operations between its factories. 

**GLOBAL MANUFACTURING SOLUTIONS** has enlisted the expertise of **IMPERIAL DYNAMICS CONSULTING**, a renowned consulting firm specializing in data analytics. 

**IMPERIAL DYNAMICS CONSULTING** will leverage their team of skilled data and business analysts to provide comprehensive data analytics consulting services. 

The goal is to gain deeper insights into the logistics and transportation processes of GLOBAL MANUFACTURING SOLUTIONS, identifying areas for improvement and optimization.

## The Assignment 

To address the business case, we have outlined **four distinct scenarios**, each addressing a specific client question. 

The class will be divided into **four groups**, and each group will be randomly assigned **one of these scenarios** to answer a unique client question. 

To facilitate the role-playing exercise, each group will be divided into **two subgroups**. 

Within each group, the subgroups will simultaneously take on the roles of the consulting team and the client. In other words, you will wear two hats: one as the client and the other as the consultants. 

**As the consulting team**, your task will be to analyze the given data and provide answers and recommendations to the client's question. 

**As the client**, you will critically evaluate the suggestions provided by the consulting team. You will ask questions to ensure the recommendations are robust and well-thought-out.

## The client

The client produces **one type of product** across **four different assembly factories** that require **four different parts** to be assembled together. That is:

$$PRODUCT = PART 1 + PART 2 + PART 3 + PART 4$$

The parts are produced in **four feeder factories** located in Hanoi, Milan, Mumbai, and Stockholm. The parts are then shipped to **four assembly factories** located in Chicago, Krakow, Munich, and Shanghai for final assembly into the client’s product.

Each feeder factory produces **only one of the four types of parts** needed in the final product, 

+ Hanoi produces PART 1,
+ Milan produces PART 2,
+ Mumbai produces PART 3,
+ Stockholm produces PART 4.

Since the assembly factories need all four types of parts to assemble the final product, each feeder factory must deliver to all the final assembly factories. For example, the feeder factory in Mumbai will need to deliver PART 3 to Chicago, Krakow, Munich, and Shanghai for final assembly.

In other words, since one of each type of part is needed for final assembly, each final assembly factory receives one part from each feeder factory. For example, the Chicago assembly factory receives Part 1 from Hanoi, Part 2 from Milan, Part 3 from Mumbai, and Part 4 from Stockholm. The other assembly factories receive parts in the same manner. The data that the client has shared captures the full picture of the movement of goods.

## The scenarios

You will be assigned to one of the four scenarios. To provide the answers, you can make assumptions. Remember, **our focus is on presentation style and skills**, and how you adapt the content of the presentation to the right audience.

**Scenario 1 – Visionary Insight**

In this scenario, the CEO of GLOBAL MANUFACTURING SOLUTIONS, a visionary leader with a keen eye on the company's strategic direction, is contemplating redistributing operations to strengthen the company's market position and capture new growth opportunities. The sales team has projected a global demand of 125 final products per month. However, the critical question remains: Does the current production capacity suffice to meet this demand? If not, what operational changes are necessary, and what would be the expected operational expenditure (OPEX) implications? The CEO is particularly focused on understanding the market dynamics and business considerations that should guide this pivotal decision.

**Scenario 2 – Foundation Building** 

In this scenario, the logistics department of GLOBAL MANUFACTURING SOLUTIONS has noticed a recurring anomaly in their operations: the feeder factory in Stockholm has reported abnormally high shipment costs. This anomaly has prompted a thorough investigation. The logistics team is interested in uncovering the root cause of this issue. Why is Stockholm's shipment cost so high? What factors are contributing to this anomaly? The logistics team is interested in learning from actionable insights to correct this issue and ensure smooth, cost-effective operations moving forward.

**Scenario 3 – Engineer’s Edge** 

In this scenario, the operations team at GLOBAL MANUFACTURING SOLUTIONS is facing a critical challenge. Several assembly factories have reported complaints about delays at the feeder factory in Mumbai. The feeder factory manager in Mumbai claims that the assembly factories are not placing orders with sufficient lead time, causing a bottleneck in the production process. The operations team requires you to dive deep into the situation to provide clarity. What is causing these delays? Is it a communication breakdown, a logistical issue, or something else? The operations team expects that you identify the root cause and propose solutions to streamline the process and eliminate delays. They are particularly interested in understanding the technical details of how you uncovered the root cause.

**Scenario 4 – Architect’s Vision** 

In this scenario, the CEO of GLOBAL MANUFACTURING SOLUTIONS, a leader with a strong technical interest, is facing mounting pressure from stakeholders to reduce the company's carbon emissions. The board has proposed a significant change: decreasing air shipments as a strategy to lower emissions. The CEO is determined to understand how exactly reducing air shipments will impact the company's carbon footprint. What methodology should be used to assess this impact accurately? The CEO seeks a comprehensive understanding of the methodology used to claim environmental impacts and the practical implications of this strategy.

## The data

### Data description


The client has shared **one month of their complete shipment data** for November 2024. The month shared represents a typical month. The data includes details about origins, destinations, order dates, dispatch dates, expected delivery dates, actual delivery dates, transportation modes, weights, and costs. For the sake of simplicity, each row corresponds to one part shipped.

**Main file:**

1. Shipment Data (Sheet: November_2024_Transportation)

This sheet contains detailed information about individual shipments, including:

* **Shipment ID:** Unique identifier for each shipment.
* **Item:** The part being shipped.
* **Order, Dispatch, Expected Delivery, and (Actual) Delivery Dates:** The date of order, dispatch, and expected and actual delivery timelines.
* **Transportation Days:** Total transit time for each shipment.
* **Origins and Destinations:** Cities where shipments start and end.
* **Weight and Distance:** Weight of goods transported (in kilograms) and the distance traveled (in kilometers).
* **Delivery Status:** Indicator of whether shipments were delivered on time. 
* **Transportation Mode:** Mode of transport (Road, Ocean, Air).
* **Fixed, Variable, and Total Costs:** Fixed, variable and total costs in GBP.

**Supporting files:**

2. Transportation Cost Data (Sheet: Transportation Cost)

This sheet provides unit costs for various transportation modes:

* **Fixed Cost (£/km):** Cost associated with distance regardless of shipment weight.
* **Variable Cost (£/kg):** Cost per kilogram of goods transported.
* **Comments:** Descriptions of the advantages and disadvantages of each transportation mode (e.g., air freight is fast but expensive).

3. Distances and Transport Preferences (Sheet: Distances)

This sheet includes:

* **Origins and Destinations:** Geographical details of the cities involved in transportation.
* **Latitude and Longitude:** Coordinates for mapping and distance calculations.
* **Country Codes:** Origin and destination countries for shipments.
* **Distances (km):** Straight-line distances between origin and destination cities.
* **Preferred Transport Modes:** Suggested modes of transport based on efficiency and cost-effectiveness.

4. City Information (Sheet: Cities)

This sheet lists:

* **City Names:** Key cities involved in the transportation network.
* **Coordinates:** Latitude and longitude for geospatial reference.
* **Country Codes:** CoLet me know if you need any further assistance!

### Data quality analysis

Importing the data:

In [2]:
import pandas as pd
from geopy.distance import geodesic

# Load all sheets into a dictionary
file_path = "November_2024_Transportation_Dataset.xlsx"
all_sheets = pd.read_excel(file_path, sheet_name=None)

shipments = all_sheets['November_2024_Transportation']
transport_cost = all_sheets['Transportation cost']
# Drop the 'Comments' column if it exists
if 'Comments' in transport_cost.columns:
    transport_cost = transport_cost.drop(columns=['Comments'])
routes = all_sheets['Distances']

cities = all_sheets['Cities']

Visualising the first rows:

In [3]:
shipments.head()

Shipment_ID    Item Order_Date Dispatch_Date Expected_Delivery_Date  \
0    SHP00001  PART 1 2024-11-01    2024-11-01             2024-11-29   
1    SHP00002  PART 2 2024-11-01    2024-11-02             2024-11-10   
2    SHP00003  PART 2 2024-11-01    2024-11-02             2024-11-10   
3    SHP00004  PART 1 2024-11-02    2024-11-02             2024-11-23   
4    SHP00005  PART 2 2024-11-02    2024-11-02             2024-11-11   

  Delivery_Date  Transportation_Days Origin Destination  Weight_kg  \
0    2024-11-29                   28  Hanoi    Shanghai        761   
1    2024-11-10                    8  Milan      Krakow       1429   
2    2024-11-10                    8  Milan      Krakow       1429   
3    2024-11-23                   21  Hanoi      Munich        761   
4    2024-11-11                    9  Milan      Munich       1429   

   Distance_km Delivery_Status Transportation_Mode  Fixed_Cost_GBP  \
0  1924.459238         On-Time               Ocean      673.560733   
1   953.414225         On-Time                Road      476.707113   
2   953.414225         On-Time                Road      476.707113   
3  8602.650679         On-Time               Ocean     3010.927738   
4   348.531624         On-Time                Road      174.265812   

   Variable_Cost_GBP  Total_Cost_GBP  
0              38.05      711.610733  
1             142.90      619.607113  
2             142.90      619.607113  
3              38.05     3048.977738  
4             142.90      317.165812

Data type and completeness:

In [4]:
shipments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Shipment_ID             500 non-null    object        
 1   Item                    500 non-null    object        
 2   Order_Date              500 non-null    datetime64[ns]
 3   Dispatch_Date           500 non-null    datetime64[ns]
 4   Expected_Delivery_Date  500 non-null    datetime64[ns]
 5   Delivery_Date           500 non-null    datetime64[ns]
 6   Transportation_Days     500 non-null    int64         
 7   Origin                  500 non-null    object        
 8   Destination             500 non-null    object        
 9   Weight_kg               500 non-null    int64         
 10  Distance_km             500 non-null    float64       
 11  Delivery_Status         500 non-null    object        
 12  Transportation_Mode     500 non-null    object    

**Note:** Lucky day! The client has provided data in standardized formats with complete information. **Keep in mind** that in real-world scenarios, this level of data quality is uncommon. Many industries, especially those operating with extensive legacy systems, rely heavily on **individual expertise** and **decentralized data sources**.

### Initial data insights:

In [5]:
shipments.describe()

Order_Date                  Dispatch_Date  \
count                            500                            500   
mean   2024-11-14 22:01:55.199999744  2024-11-19 09:44:38.399999744   
min              2024-11-01 00:00:00            2024-11-01 00:00:00   
25%              2024-11-07 00:00:00            2024-11-14 00:00:00   
50%              2024-11-14 00:00:00            2024-11-21 00:00:00   
75%              2024-11-23 00:00:00            2024-11-26 00:00:00   
max              2024-11-30 00:00:00            2024-11-30 00:00:00   
std                              NaN                            NaN   

              Expected_Delivery_Date        Delivery_Date  \
count                            500                  500   
mean   2024-11-24 15:01:26.400000256  2024-11-24 22:33:36   
min              2024-11-05 00:00:00  2024-11-06 00:00:00   
25%              2024-11-20 00:00:00  2024-11-20 00:00:00   
50%              2024-11-26 00:00:00  2024-11-26 00:00:00   
75%              2024-11-29 06:00:00  2024-11-30 00:00:00   
max              2024-12-06 00:00:00  2024-12-15 00:00:00   
std                              NaN                  NaN   

       Transportation_Days   Weight_kg   Distance_km  Fixed_Cost_GBP  \
count           500.000000   500.00000    500.000000      500.000000   
mean              5.534000   925.61800   6040.739670     3031.764881   
min               0.000000   246.00000    348.531624      174.265812   
25%               0.000000   761.00000   1315.887532      778.704514   
50%               6.000000  1201.00000   6899.752871     2785.410742   
75%               7.000000  1429.00000   8602.650679     4537.979476   
max              36.000000  1429.00000  12965.655647     9724.241735   
std               4.917031   442.56846   4102.337117     2243.467309   

       Variable_Cost_GBP  Total_Cost_GBP  
count         500.000000      500.000000  
mean          157.529900     3189.294781  
min            12.300000      317.165812  
25%            38.050000     1139.004514  
50%            76.100000     2823.460742  
75%           228.300000     4554.221174  
max           428.700000     9905.809659  
std           144.478309     2263.608134

* **Order Dates:** The data spans the month of November 2024, with orders ranging from the 1st to the 30th. The median order date is November 14.
* **Dispatch Dates:** The fastest dispatch occurred on the same day as the order (November 1).
* **Expected vs. Actual Delivery:** The earliest delivery occurred on November 6, while the earliest expected delivery was November 5.
* **Transportation Days:** The average transit time is approximately 5.5 days. The shortest shipments were delivered on the same day, while the longest took 36 days.
* **Weight of Shipments:** The average shipment weight is approximately 925 kg, ranging from a light shipment of 246 kg to a heavier one of 1,429 kg.
* **Distances:** The average distance traveled by shipments is 6,041 km, with a wide range—from around 349 km to over 12,965 km.
* **Fixed Costs:** The average fixed cost is around £3,031, ranging from approximately £174 to £9,724.
* **Variable Costs:** The average variable cost is around £157, ranging from approximately £12 to £428.
* **Total Costs:** The average total cost is around £3,189, ranging from approximately £317 to £9,905.

### Bonus formulas - Not Needed for the assigment

The distances between cities have already been provided by the client. However, if this information were not available, the following formulas could come in handy:

**1)** Calculating distances between cities with the library **geopy**, and the function **geodesic**

```python
from geopy.distance import geodesic

def calculate_distance(row):
    origin_coords = (row["Origin_Lat"], row["Origin_Lng"])
    destination_coords = (row["Destination_Lat"], row["Destination_Lng"])
    return geodesic(origin_coords, destination_coords).kilometers
    
routes["Distance_km"] = routes.apply(calculate_distance, axis=1)
```
**2)** Getting coordinates using the name of cities with the library **geopy**, and the function **Nominatim**

```python
from geopy.geocoders import Nominatim

def get_coordinates(city_name):
    geolocator = Nominatim(user_agent="shipmentsIDlocations")
    location = geolocator.geocode(city_name)
    return location.latitude, location.longitude

# Add latitude and longitude for cities
cities['Lat'], cities['Lon'] = zip(*cities['City'].apply(get_coordinates))

### Visualising the shipments

Route visualisation - world map:

In [9]:
import plotly.graph_objects as go
import plotly.io as pio
pio.renderers.default = "vscode"

# Get the coordinated in the same dataframe
shipments_coordinates = pd.merge(shipments, routes[['Origin', 'Destination', 'Origin_Lat', 'Origin_Lon','Destination_Lat', 'Destination_Lon']],on=['Origin', 'Destination'])

# Create the map
fig = go.Figure()

# Add markers for Origins
fig.add_trace(go.Scattergeo(
    locationmode='country names',
    lon=shipments_coordinates['Origin_Lon'],
    lat=shipments_coordinates['Origin_Lat'],
    mode='markers',
    marker=dict(size=8, color='darkblue', symbol='circle'),  # Blue circles for origins
    name='Origin',
    text=shipments_coordinates['Origin']  # City names as hover text
))

# Add markers for Destinations
fig.add_trace(go.Scattergeo(
    locationmode='country names',
    lon=shipments_coordinates['Destination_Lon'],
    lat=shipments_coordinates['Destination_Lat'],
    mode='markers',
    marker=dict(size=8, color='lightcoral', symbol='diamond'),  # Red diamonds for destinations
    name='Destination',
    text=shipments_coordinates['Destination']  # City names as hover text
))

# Add lines between Origins and Destinations
for i in range(len(shipments_coordinates)):
    fig.add_trace(go.Scattergeo(
        locationmode='country names',
        lon=[shipments_coordinates['Origin_Lon'][i], shipments_coordinates['Destination_Lon'][i]],
        lat=[shipments_coordinates['Origin_Lat'][i], shipments_coordinates['Destination_Lat'][i]],
        mode='lines',
        line=dict(width=2, color='gray'),
        name=f"{shipments_coordinates['Origin'][i]} → {shipments_coordinates['Destination'][i]}"
    ))

# Update map layout
fig.update_layout(
    title='Shipments: Origins and Destinations',
    geo=dict(
        scope='world',
        projection_type='natural earth',
        showland=True,
        landcolor='rgb(243, 243, 243)',
        coastlinecolor='rgb(204, 204, 204)'
    )
)

# Display the map
fig.show()

Route visualisation - Sankey diagram:

In [10]:
def create_sankey(data, title):
    # Step 1: Count shipments for each Origin-Destination pair
    grouped_df = data.groupby(['Origin', 'Destination', 'Item']).size().reset_index(name='Value')
    
    # Step 2: Create unique labels for Sankey nodes
    all_nodes = list(pd.concat([grouped_df['Origin'], grouped_df['Destination']]).unique())
    node_indices = {node: idx for idx, node in enumerate(all_nodes)}
    
    item_colors = {
        'PART 1': 'dimgray',    # Dark gray for Product 1
        'PART 2': 'darkgray',        # Medium gray for Product 2
        'PART 3': 'silver',   # Light gray for Product 3
        'PART 4': 'gainsboro'    # Very light gray for Product 4
    }
    
    
    # Step 3: Assign colors based on the item
    grouped_df['Color'] = grouped_df['Item'].map(item_colors)
    
    # Step 4: Create unique labels for Sankey nodes
    all_nodes = list(pd.concat([grouped_df['Origin'], grouped_df['Destination']]).unique())
    node_indices = {node: idx for idx, node in enumerate(all_nodes)}
    
    # Step 5: Prepare data for Sankey diagram
    sankey_data = {
        'node': {
            'label': all_nodes,  # Labels for nodes
            'color': ['darkblue' if node in grouped_df['Origin'].unique() else 'lightcoral' for node in all_nodes]
      
        },
        'link': {
            'source': grouped_df['Origin'].map(node_indices),  # Map origins to indices
            'target': grouped_df['Destination'].map(node_indices),  # Map destinations to indices
            'value': grouped_df['Value'],  # Shipment count as link value
            'color': grouped_df['Color'],  # Link colors based on items
            'label': grouped_df['Item']
        }
    }
    
    # Step 6: Create the Sankey diagram
    fig = go.Figure(go.Sankey(
        node=dict(
            pad=15,
            thickness=20,
            line=dict(color="black", width=0.5),
            label=sankey_data['node']['label'],
            color=sankey_data['node']['color']
        ),
        link=dict(
            source=sankey_data['link']['source'],
            target=sankey_data['link']['target'],
            value=sankey_data['link']['value'],
            color=sankey_data['link']['color'],
            label=sankey_data['link']['label'],
        )
    ))
    # Update layout
    return fig.update_layout(title_text=title, font_size=12)

fig = create_sankey(shipments_coordinates, "Origin-Destination Sankey Diagram")
# Show the diagram
fig.show()

## Bonus - Useful Tips for Tackling Problems

When approaching complex problems, it’s tempting to aim for a full-scale optimization of the entire system. While this goal is appealing, the reality is that constraints like budgets, time, and resources often make it impractical. Instead, focusing on specific, impactful areas can yield significant improvements and demonstrate value.

Some tips and frameworks to guide your approach:

### Problem Bounding

#### 1. Focus on the Vital Few: The Pareto Principle

The Pareto Principle, also known as the 80/20 Rule, suggests that 80% of the outcomes are often driven by 20% of the inputs. This principle helps narrow the focus to areas that contribute most to the problem:

Examples in logistics:
* 80% of transportation costs may arise from just 20% of the routes.
* 80% of delays may occur due to inefficiencies in 20% of the factories or transportation modes.

How to apply this in your analysis:
Use the data to identify the "vital few" factors driving the majority of the issues, such as costliest routes or consistently delayed factories. Prioritize these high-impact areas for optimization.

#### 2. Prioritize "Low-Hanging Fruits" Within the Prioritized Areas

While the Pareto Principle helps identify the key areas to focus on, it’s essential to consider low-hanging fruits within those areas. These are opportunities where small, straightforward changes can yield significant results. This is particularly true when considering the environment in which each factory operates.

Why focus on low-hanging fruits?
* They demonstrate quick wins, which build credibility and momentum.
* They often require fewer investments.

How to identify low-hanging fruits? Look for patterns in the data where inefficiencies are clear and addressable, such as:
* Routes with consistently high costs.
* Factories with frequent delays.
* Over-reliance on expensive transportation modes for non-urgent shipments.

In a real environment, you would also consider the availability of local resources, prior knowledge of the factory, level of local collaboration, etc.

By targeting low-hanging fruits in high-priority areas, you create immediate value while setting the stage for more comprehensive optimization.

### Problem Resolution

#### 1. Root Cause Analysis or Ishikawa (Fishbone) Diagram

What is the Ishikawa Diagram?
Also known as the "fishbone" diagram, this tool helps systematically identify the root causes of a problem by categorizing potential factors.

Uses the 4 Ps or 4 Ms (sometimes 6 Ms):
* 4 Ps when analyzing service-related or organizational issues.
* 4 Ms for manufacturing, production, or logistics-related problems.

**The Four Ps** Primarily used in service industries, the 4 Ps categorize the causes of problems into:

* **P**eople: Human-related factors, such as skills, training, attitude, or communication.
* **P**rocesses: Workflow and procedural inefficiencies or inconsistencies.
* **P**olicies: Rules, regulations, or management decisions impacting operations.
* **P**lant/Place: Physical location or environment where the process occurs.

**The Four Ms (or Six Ms)** Primarily used in manufacturing industries:

* hu**M**an (People): Human factors, including workforce skills, training, and engagement.
* **M**achine: Equipment, tools, and technology issues.
* **M**ethod (Process): The way tasks or processes are performed.
* **M**aterials: The raw materials or inputs used in production.

**Extension:**
* **M**easurement: Metrics and measurements, including incorrect or insufficient data.
* **M**other Nature (Environment): Environmental factors, including weather, climate, or natural conditions.

How to use Root Cause Analysis:
* Start with a problem (e.g., delays at a specific factory).
* Break it down into contributing categories such as:
    * People: Staff inefficiencies or skills gaps.
    * Processes: Poor scheduling or lack of standardization.
    * Equipment: Transport vehicle issues or capacity constraints.
    * Environment: Weather, road conditions, or port congestion.

Dive deeper into each category to pinpoint actionable root causes. This system ensures that your solutions target the underlying issues rather than just addressing symptoms.